netket/machine/torch.py

```diff
@@ -18,12 +24,27 @@ class Torch(AbstractMachine):
     def __init__(self, module, hilbert):
         self._module = _torch.jit.load(module) if isinstance(module, str) else module
         self._module.double()
+
+        if _backpack_imported:
+            self._use_backpack = True
+            for module in self._module.children():
+                if not isinstance(module, (_torch.nn.Conv2d, _torch.nn.Linear)) \
+                   and sum(p.numel() for p in module.parameters()):
```

Comment:
That doesn't look right...

First of all, Backpack expects a module to be `torch.nn.Sequential` (see documentation). I.e. you should do `isinstance(module, torch.nn.Sequential)` for normal modules and `module.original_name == Sequential` (or something similar) for `torch.jit.ScriptModule`.

Secondly, not all parameter-less layers are supported by Backpack (see documentation). You should probably use an explicit list of supported layers or just wrap `extend` into a `try: ... except: ...` block.

Finally, there's the requirement that `forward` function should not be overridden (see documentation). This is non-trivial to check automatically. E.g. you could check whether the module *is* `torch.nn.Sequential` is or derived from it etc. But this doesn't cover the case when the user simply assigns a `forward` method dynamically... This will also break for `ScriptModule`s.

All in all, I'd suggest dropping the idea of automatically detecting when Backpack can be used and simply introduce a flag `use_backpack` and delegate the decision to the user who knows the network architecture.

Reply:
Thanks for your reply. 

These constraints are necessary for backpack to work in both 1st-order extensions (jacobian included) and some other 2nd-order extensions. In my tests backpack works even we write a forward function or add some parameter-less layers created by ourselves or don't write it as a sequence. Otherwise some necessary manipulations like periodic boundaries will be impossible.

But I agree that the use of backpack in this way is not 100% safe. Change it to a manual opt-in may a better option.

Reply:
OK, for me the important thing is that this is optional and that we don't use it when it doesn't work ! If it is hard to check automatically whether it is safe to use bacpack, then I agree with @twesterhout that the best solution is just to put a ```use_backpack``` argument in the ```__init__``` of the Torch class, and leave the user the burden to check if this makes sense for the model chosen